### PR TITLE
Issue 14 rcomma non en locales

### PR DESCRIPTION
**Pull request by mpyat2**
Hi David @dbenn,

This branch addresses issues #14 and #92.
The following files/classes have been changed:
ApacheCommonsLoessFitter.java
ApacheCommonsPolynomialFitCreatorPlugin.java
PeriodAnalysisDerivedMultiPeriodicModel.java
PeriodFitParameters.java

toRString(): locale-independent formatting (using ENGLISH locale)
toExcelString(): formula separator should be locale-dependent. If ',' is a decimal separator, the Excel formula separator is changed to ';'.

ApacheCommonsLoessFitter, ApacheCommonsPolynomialFitCreatorPlugin: polynomial coefficients are formatted now using scientific notation (with configured precision: in this case it is a total number of significant digits) via new functions NumericPrecisionPrefs.formatPolyCoef/formatPolyCoefLocaleIndependent. This is because the coefficients could have a huge range and must be formatted with exponential notation (at least in some cases).

Before the fix, polynomial coefficients were converted to strings implicitly (with implicit Double.toString() call), so they did not respect locale at all.

Excel formula issue (locale-dependent list separator) arose because the Excel formulas are implemented with SUM() function. We cannot use ',' when the decimal separator is also a comma (in my locale with ',' as a decimal separator, the default list separator is ';'). Why not get rid of SUM(a, b, c ...)? The formula could be in the form a + b + c +..., like other ones. What do you think?

Apparently, there is also a problem with functions for models generated by Loess plug-in. First, they are too long. And it seems they are incorrect. Maybe I'm doing something wrong, yet I cannot regenerate the model with those functions...

Concerning #138: formatPolyCoef returns string representation of a number in scientific notation with E+replaced by E as a workaround.

### CODE REVIEW

File: src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java [hunk 1]
```diff
@@ -156,10 +187,95 @@ public static DecimalFormat getOtherOutputFormat() {
...
+			s += ".0";
+		}
+		return s;
+	}
+*/	
+	
+	// Excel formula separator (list separator)
+	// There is no way to get system "list separator" (Windows specific
+	// locale-sensitive parameter). So we make an assumption. 
+	// Why not replace SUM() in Excel formulas with a simple a+b+c... expression?
```
Comment (dbenn) [inline src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java:267]: this seems like a reasonable approach @mpyat2

File: src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java [hunk 2]
```diff
@@ -156,10 +187,95 @@ public static DecimalFormat getOtherOutputFormat() {
...
+	public static String formatPolyCoefLocaleIndependent(double num) {
+		//return formatGeneralLocaleIndependent(num);
+		return formatScientificLocaleIndependent(num);
+	}
+	
+	// Scientific format
+
+	private static String formatScientific(double num) {
+		// .replace("E+", "E"): VeLa issue workaround
+		return String.format(Locale.getDefault(), "%." + otherDecimalPlaces + "E", num).replace("E+", "E");
```
Comment (dbenn) [inline src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java:229]: a reasonable workaround until I fix #138 @mpyat2

File: src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java [hunk 3]
```diff
@@ -156,10 +187,95 @@ public static DecimalFormat getOtherOutputFormat() {
...
 	}
 
+	// PMAK (2021-03-25): 
+	// Special case: format for polynomial coefficients.
+	// These coefficients have a huge value range and cannot be satisfactorily
+	// represented by DecimalFormat.
+	// What format would be better: general (%.xxG) or pure scientific (%.xxE)?
+	// Both have issues (for the current VeLa numeric parser):
+	//   1) signed positive exponent cannot be correctly parsed by VeLa now (i.e. 1.234E+15)
+	//   2) long number strings without decimal separator cannot be parsed too (i.e. -608516245008941)
```
Comment (dbenn) [inline src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java:209]: @mpyat2 Thanks. I'm adding NumberParser unit tests to look at this. I'll either address it in #138 or in a separate issue. VeLa integers are currently Java ints. Should be longs. For now, the workaround would be `-608516245008941.0` to force VeLa to parse it as a real.

File: plugin/src/org/aavso/tools/vstar/external/plugin/ApacheCommonsLoessFitter.java [hunk 4]
```diff
@@ -139,15 +140,16 @@ public String toString() {
...
 						double constCoeff = 0;
 
+						// We use formatGeneral because of a huge range
 						for (PolynomialFunction f : function.getPolynomials()) {
 							double[] coeffs = f.getCoefficients();
 							for (int i = coeffs.length - 1; i >= 1; i--) {
-								strRepr += "    " + coeffs[i];
-								strRepr += "t^" + i + "+\n";
+								strRepr += "    " + NumericPrecisionPrefs.formatPolyCoef(coeffs[i]);
+								strRepr += "*t^" + i + "+\n";
```
Comment (dbenn) [inline plugin/src/org/aavso/tools/vstar/external/plugin/ApacheCommonsLoessFitter.java:148]: Thanks @mpyat2. This was a hang-over from when functions were not VeLa, merely "descriptive" and I obviously did not test this transition properly for Loess models.

**Review by dbenn — COMMENTED** [review]
Thanks for your work on this so far. See my specific comments re: code lines.

File: src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java (hunk 3 above)
Comment (mpyat2) [reply src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java:209]: ok

File: src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java (hunk 2 above)
Comment (mpyat2) [reply src/org/aavso/tools/vstar/util/prefs/NumericPrecisionPrefs.java:229]: ok

**Comment by mpyat2** [comment]
I've created issue #145 for the ApacheCommonsLoessFitter model problem

**Comment by mpyat2** [comment]
In the last commit, I've removed Excel =SUM() function and got rid of Excel list separator (not used anymore).
I've also replaced incorrect ApacheCommonLoessFitter model functions by Mediator.NOT_IMPLEMENTED_YET text.
A new issue  #145 for the ApacheCommonsLoessFitter has been created.

**Comment by dbenn** [comment]
@mpyat2 This one looks close. Anything remaining from your POV? If not, we could merge.

**Comment by mpyat2** [comment]
@dbenn Yes, please merge

**Pull request merged**